src/dmd/compiler.d

```diff
@@ -107,38 +111,54 @@ struct Compiler
      */
     extern (C++) static Expression paintAsType(Expression e, Type type)
     {
-        // We support up to 512-bit values.
-        ubyte[64] buffer;
+        union U
+        {
+            d_int32 int32value;
+            d_int64 int64value;
+            float float32value;
+            double float64value;
+        }
+        U u = void;
+
         assert(e.type.size() == type.size());
-        // Write the expression into the buffer.
+
         switch (e.type.ty)
         {
         case Tint32:
         case Tuns32:
+            u.int32value = cast(d_int32) e.toInteger();
+            break;
         case Tint64:
         case Tuns64:
-            encodeInteger(e, buffer.ptr);
+            u.int64value = cast(d_int64) e.toInteger();
             break;
         case Tfloat32:
+            u.float32value = cast(float) e.toReal();
+            break;
         case Tfloat64:
-            encodeReal(e, buffer.ptr);
+            u.float64value = cast(double) e.toReal();
             break;
         default:
-            assert(0);
+            assert(0, "Unsupported source type");
         }
-        // Interpret the buffer as a new type.
+
+        real_t r = void;
         switch (type.ty)
         {
         case Tint32:
         case Tuns32:
+            return new IntegerExp(e.loc, u.int32value, type);
```

Comment:
Should think about doing this in a way that doesn't allocate.

Reply:
No change from the existing implementation...